include/ringmesh/basic/attributes.h

```diff
@@ -0,0 +1,1024 @@
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <geogram/basic/geofile.h>
+#include <map>
+#include <ringmesh/basic/common.h>
+#include <ringmesh/basic/factory.h>
+#include <ringmesh/basic/logger.h>
+#include <ringmesh/basic/logger.h>
+#include <set>
+#include <typeinfo>
+
+/**
+ * \file ringmesh/basic/attributes.h
+ * \brief Generic mechanism for attributes.
+ * This file has been strongly inspired from the attributes of the geogram
+ * library.
+ * Many thanks to Bruno Levy (Bruno.Levy@inria.fr) who did the first
+ * implementation in Geogram.
+ */
+
+namespace RINGMesh
+{
+    class AttributeStore;
+
+    class RINGMESH_API Store
+    {
+    public:
+        Store( index_t elemsize ) : element_size_( elemsize )
+        {
+        }
+
+        virtual ~Store() = default;
+        /**
+        * \brief Gets the size.
+        * \return the number of items
+        */
+        index_t size() const
+        {
+            return size_;
+        }
+
+        /**
+        * \brief Gets the total number of elements.
+        * \details This corresponds to one position past
+        *  the last valid index.
+        * \return the total number of elements.
+        */
+        index_t nb_elements() const
+        {
+            return size_;
+        }
+        virtual const void* data() const = 0;
+        virtual void* data() = 0;
+
+        virtual const void* element( index_t element ) const = 0;
+        virtual void* element( index_t element ) = 0;
+
+        virtual void compress( const std::vector< index_t >& old2new ) = 0;
+        /**
+        * \brief Applies a permutation to the stored attributes.
+        * \details Applying a permutation to the data is equivalent
+        *  to:
+        * \code
+        * for(i=0; i<permutation.size(); i++) {
+        *    data2[i] = data[permutation[i]]
+        * }
+        * data = data2 ;
+        * \endcode
+        * But it is done in-place.
+        * \param[in] permutation the permutation.
+        *  It is temporarily changed during execution of the
+        *  function, but identical to the input on exit.
+        * \note This function uses memcpy(). If required, it
+        *  can be overloaded in derived classes.
+        */
+        virtual void apply_permutation(
+            const std::vector< index_t >& permutation ) = 0;
+
+        virtual bool elements_type_matches(
+            const std::string& type_name ) const = 0;
+        virtual std::string element_typeid_name() const = 0;
+
+        virtual void resize( index_t new_size ) = 0;
+
+        virtual void clear() = 0;
+
+        virtual Store* clone() const = 0;
+
+        index_t element_size() const
+        {
+            return element_size_;
+        }
+
+    protected:
+        index_t size_{ 0 };
+        index_t element_size_; // size of element in byte

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)